subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c

```diff
@@ -0,0 +1,761 @@
+
+	instance = find_index(obj_inst_id);
+
+	return handle_event(instance, EVENT_DEACTIVATE);
+}
+
+static int package_write_cb(uint16_t obj_inst_id, uint16_t res_id, uint16_t res_inst_id,
+			    uint8_t *data, uint16_t data_len, bool last_block, size_t total_size)
+{
+	int ret = -EINVAL;
+	struct lwm2m_swmgmt_data *instance = NULL;
+
+	instance = find_index(obj_inst_id);
+
+	ret = handle_event(instance, EVENT_PKG_URI_WRITE);
+
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = instance->write_package_cb(obj_inst_id, res_id, res_inst_id, data, data_len,
+					 last_block, total_size);
+
+	if (ret < 0) {
+		handle_event(instance, EVENT_DOWNLOAD_FAILED);
+		switch (ret) {
+		case -ENOMEM:
+			instance->update_result = UPD_RES_OUT_OF_MEMORY_DURING_DOWNLOAD;
+			break;
+		case -ENOSPC:
+			instance->update_result = UPD_RES_NOT_ENOUGH_STORAGE;
+			ret = -EFBIG;
+			break;
+		case -EFAULT:
+			instance->update_result = UPD_RES_PACKAGE_INTEGRITY_CHECK_FAILURE;
+			break;
+		default:
+			instance->update_result = UPD_RES_LOST_CONNECTION_DURING_DOWNLOAD;
+			break;
+		}
+
+		return ret;
+	}
+
+	if (last_block) {
+		handle_event(instance, EVENT_PKG_WRITTEN);
+	}
+
+	return 0;
+}
+
+static void set_update_result(uint16_t obj_inst_id, int error_code)
+{
+	int ret;
+	struct lwm2m_swmgmt_data *instance;
+
+	instance = find_index(obj_inst_id);
+
+	if (error_code == 0) {
+		handle_event(instance, EVENT_PKG_WRITTEN);
+
+		/* If the verify function wasn't provided, skip the check. */
+		if (verify_package) {
+			ret = verify_package();
+		} else {
+			ret = 0;
+		}
+
+		if (ret == 0) {
+			handle_event(instance, EVENT_PKG_INTEGRITY_VERIFIED);
+		} else {
+			handle_event(instance, EVENT_PKG_INTEGRITY_FAILED);
+		}
+
+		return;
+	}
+
+	handle_event(instance, EVENT_DOWNLOAD_FAILED);
+	if (error_code == -ENOMEM) {
+		instance->update_result = UPD_RES_OUT_OF_MEMORY_DURING_DOWNLOAD;
+	} else if (error_code == -ENOSPC) {
+		instance->update_result = UPD_RES_NOT_ENOUGH_STORAGE;
+	} else if (error_code == -EFAULT) {
+		instance->update_result = UPD_RES_PACKAGE_INTEGRITY_CHECK_FAILURE;
+	} else if (error_code == -ENOTSUP) {
+		instance->update_result = UPD_RES_INVALID_URI;
+	} else {
+		instance->update_result = UPD_RES_LOST_CONNECTION_DURING_DOWNLOAD;
+	}
+}
+
+static int package_uri_write_cb(uint16_t obj_inst_id, uint16_t res_id, uint16_t res_inst_id,
+				uint8_t *data, uint16_t data_len, bool last_block,
+				size_t total_size)
+{
+#ifdef CONFIG_LWM2M_FIRMWARE_UPDATE_PULL_SUPPORT
+	int error_code;
+	struct lwm2m_swmgmt_data *instance = NULL;
+
+	instance = find_index(obj_inst_id);
+
+	struct requesting_object req = { .obj_inst_id = obj_inst_id,
+					 .is_firmware_uri = false,
+					 .result_cb = set_update_result,
+					 .write_cb = instance->write_package_cb,
+					 .verify_cb = NULL };
+
+	verify_package = req.verify_cb;
+
+	error_code = lwm2m_pull_context_start_transfer(instance->package_uri, req, K_NO_WAIT);
+
+	if (error_code) {
+		return error_code;
```

Comment:
Not sure here, this will result in an error response being sent to the server if some other download is in progress, wouldn't that be confusing for the server side (all in all URI write would succeed).

Shouldn't we rather have an extra delayed work item (per instance) and set it to retry download after some time in case of `EALREADY`?

Reply:
That's a good question... should we? Adding a delayed work would bump up the memory usage a bit for sure. Not sure about this one either.

Reply:
I don't have a strong opinion here, I guess we coudl leave it as it is now (i. e. report an error to the server), and in cases it turns out problematic update in the future.